File: src/test/scala/com/github/mauricio/postgresql/parsers/ParserSSpec.scala
package com.github.mauricio.postgresql.parsers

import org.specs2.mutable.Specification
import org.jboss.netty.buffer.ChannelBuffers
import java.nio.charset.Charset
import com.github.mauricio.postgresql.messages.backend.{ParameterStatusMessage, Message}
import org.jboss.netty.util.CharsetUtil

/**
 * User: Maurício Linhares
 * Date: 2/28/12
 * Time: 9:59 PM
 */

class ParserSSpec extends Specification {

  val parser = new ParameterStatusParser(CharsetUtil.UTF_8)

  "ParameterStatusParser" should {

    "correctly parse a config pair" in {

      val key  = "application-name"
      val value = "my-cool-application"

      val buffer = ChannelBuffers.dynamicBuffer()

      buffer.writeBytes( key.getBytes( Charset.forName("UTF-8") ) )
      buffer.writeByte(0)
      buffer.writeBytes( value.getBytes( Charset.forName("UTF-8") ) )
      buffer.writeByte(0)

      val content = this.parser.parseMessage( buffer ).asInstanceOf[ParameterStatusMessage]

      List(
        content.key === key,
        content.value === value,
        content.name === Message.ParameterStatus,
        buffer.readerIndex() === buffer.writerIndex() )
    }

  }

}

File: postgresql-async/src/test/java/com/github/aysnc/sql/db/postgresql/parsers/ParserSSpec.kt
package com.github.aysnc.sql.db.postgresql.parsers

import com.github.jasync.sql.db.postgresql.messages.backend.ParameterStatusMessage
import com.github.jasync.sql.db.postgresql.messages.backend.ServerMessage
import com.github.jasync.sql.db.postgresql.parsers.ParameterStatusParser
import io.netty.buffer.Unpooled
import io.netty.util.CharsetUtil
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.nio.charset.Charset

class ParserSSpec {

    val parser = ParameterStatusParser(CharsetUtil.UTF_8)

    @Test
    fun `ParameterStatusParser should correctly parse a config pair`() {
        val key = "application-name"
        val value = "my-cool-application"

        val buffer = Unpooled.buffer()

        buffer.writeBytes(key.toByteArray(Charset.forName("UTF-8")))
        buffer.writeByte(0)
        buffer.writeBytes(value.toByteArray(Charset.forName("UTF-8")))
        buffer.writeByte(0)

        val content = this.parser.parseMessage(buffer) as ParameterStatusMessage

        assertThat(content.key).isEqualTo(key)
        assertThat(content.value).isEqualTo(value)
        assertThat(content.kind).isEqualTo(ServerMessage.ParameterStatus)
        assertThat(buffer.readerIndex()).isEqualTo(buffer.writerIndex())
    }
}
